### PR TITLE
Ci/william/add xref escript

### PR DESCRIPTION
**Pull request by qzhuyan**
Add a xref escript to scan all released modules/libs.

cleaning of known Xref warning will be done in follow-ups.

### CODE REVIEW

File: scripts/xref_check.eterm [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+          {erldis,mget,2},
+          {erldis,set,3},
+          {erldis_client,connect,0},
+          {et_collector,iterate,5},
+          {et_collector,report,2},
+          {et_selector,parse_event,2},
+          {et_viewer,get_collector_pid,1},
+          {et_viewer,start_link,1},
+          {eunit,test,1},
+          {ftp,start_service,1},
```
Comment (qzhuyan) [inline scripts/xref_check.eterm:288]: maybe we need `ftp` application?

File: scripts/xref-check.escript [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env escript
+
+%%% @doc
+%%% Unlike rebar3 xref check, this script runs the full xref checks in the EMQX release dir,
+%%% meaning all the modules for release are analysed.
+%%% For behavior configuration, all rebar3 related modules attributes, filters are not used in this script,
+%%% instead all the filters, checks are defined in `xref_check.eterm`
+main(_) ->
+    {ok, [Jobs]} = file:consult("scripts/xref_check.eterm"),
+    lists:foreach(fun(#{ name := Name
```
Comment (ieQu1) [inline scripts/xref-check.escript:10]: Nit: maybe factor out to a top-level function?

**Review by ieQu1 — APPROVED** [review]

File: scripts/xref-check.escript [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                       , analysis := Analysis
+                       , excl_apps := ExclApps
+                       , excl_mods := ExclMods
+                       , filters := Filters
+                       }) ->
+                          xref:start(Name),
+                          Tid = ets:new(Name, [ordered_set, named_table]),
+                          xref:set_default(Name, [{verbose,false}, {warnings,false}]),
+                          Profile = case filelib:is_file("EMQX_ENTERPRISE") of
+                                        true -> emqx-ee;
```
Comment (thalesmg) [inline scripts/xref-check.escript:20]:
```suggestion
                                        true -> 'emqx-ee';
```

File: scripts/xref-check.escript [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                          ],
+
+                          [case xref:remove_module(Name, M) of
+                               ok -> ok;
+                               %% but in doc it should return '{error, module(), Reason}`
+                               {error, xref_base, {no_such_module, M}} -> ok
+                           end || M <- ExclMods
+                          ],
+                          ModuleInfos = xref:info(Name, modules),
+                          LibInfos = xref:info(Name, modules),
```
Comment (thalesmg) [inline scripts/xref-check.escript:39]: Aren't those two values the same?

File: scripts/xref-check.escript [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                          true = ets:insert(Tid, ModuleInfos ++ LibInfos),
+                          {ok, Res0} = xref:analyse(Name, Analysis),
+                          Res = Res0 -- Filters,
+                          Res =/= [] andalso
+                              begin
+                                  put(is_warn_found, true),
+                                  io:format("** Warnings for ~p~n : ~p~n", [Name, Res])
```
Comment (thalesmg) [inline scripts/xref-check.escript:46]: Nit: maybe print the results in a new line, so the results are wider?

File: scripts/xref_check.eterm [hunk 6]
```diff
@@ -0,0 +1,536 @@
+[ %% Check undefined_function_calls
```
Comment (thalesmg) [inline scripts/xref_check.eterm:1]: Nit:

```suggestion
%% -*- mode: erlang; -*-

[ %% Check undefined_function_calls
```

File: scripts/xref_check.eterm [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+         , release_handler
+         , cuttlefish_rebar_plugin
+         ]
+   , filters =>
+         [{{coap_client,channel_apply,3},{coap_dtls_socket,close,1}},
+          {{coap_client,channel_apply,3},{coap_dtls_socket,connect,2}},
+          {{coap_client,channel_apply,3},{coap_udp_socket,close,1}},
+          {{coap_client,channel_apply,3},{coap_udp_socket,get_channel,2}},
+          {{coap_client,channel_apply,3},{coap_udp_socket,start_link,0}},
+          {{coap_client,test,0},{eunit,test,1}},
```
Comment (thalesmg) [inline scripts/xref_check.eterm:23]: Q: are eunit/test functions in the release then?  :thinking:

File: scripts/xref_check.eterm [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+         ]
+   , excl_mods =>
+         [ systools
+         , systools_make
+         , release_handler
+         , systools_relup
+         , cuttlefish_rebar_plugin
+         ]
+   , filters =>
+         [{'Elixir.Atom',to_string,1},
```
Comment (thalesmg) [inline scripts/xref_check.eterm:250]: Those functions are absent only on non-Elixir releases, right?  Maybe they should be checked for in the Elixir release?

File: scripts/xref-check.escript (hunk 4 above)
Comment (qzhuyan) [reply scripts/xref-check.escript:39]: right! will fix.

File: scripts/xref_check.eterm (hunk 8 above)
Comment (qzhuyan) [reply scripts/xref_check.eterm:251]: yes. should be part of the CI in Elixir build I guess. Is there a flag to check if it is Elixir runtime?

File: scripts/xref-check.escript (hunk 5 above)
Comment (qzhuyan) [reply scripts/xref-check.escript]: It is in new line already. 
The output is already readable to me.

example: 
```
===> Running cross reference analysis...
** Warnings for undefined_functions
 : [{'Elixir.Poison','decode!',1},
    {'Elixir.Poison','encode!',1},
    {'Elixir.Poison.EncodeError',exception,1},
    {'Elixir.Poison.Encoder.Atom',encode,2},
    {'Elixir.Poison.Encoder.BitString',encode,2},
    {'Elixir.Poison.Encoder.Float',encode,2},
    {'Elixir.Poison.Encoder.Integer',encode,2}]
make: *** [Makefile:119: xref] Error 1
```

File: scripts/xref_check.eterm (hunk 7 above)
Comment (qzhuyan) [reply scripts/xref_check.eterm:24]: I guess so if xref say so.

File: scripts/xref-check.escript (hunk 5 above)
Comment (thalesmg) [reply scripts/xref-check.escript]: Ooops, missed that newline, ignore me.

File: scripts/xref_check.eterm (hunk 8 above)
Comment (thalesmg) [reply scripts/xref_check.eterm:251]: There's a variable that's set in `emqx_vars` (`IS_ELIXIR` I think), but that should appear only in 5.0 :thinking:

File: scripts/xref_check.eterm (hunk 7 above)
Comment (thalesmg) [reply scripts/xref_check.eterm:24]: :hear_no_evil:

**Pull request merged**